Libplanet/Store/DefaultStore.cs

```diff
@@ -246,174 +271,237 @@ public override IEnumerable<TxId> IterateStagedTransactionIds()
+                    }
+
+                    yield return txid;
+                }
+            }
         }
 
         /// <inheritdoc/>
         public override Transaction<T> GetTransaction<T>(TxId txid)
         {
-            string fileId = TxFileId(txid);
-            byte[] bytes;
-
-            _txLock.EnterUpgradeableReadLock();
-            try
+            if (_txCache.TryGetValue(txid, out object cachedTx))
             {
-                LiteFileInfo file = _db.FileStorage.FindById(fileId);
-                if (file is null)
-                {
-                    return null;
-                }
-
-                using (var stream = new MemoryStream())
-                {
-                    DownloadFile(file, stream);
+                return (Transaction<T>)cachedTx;
+            }
 
-                    bytes = stream.ToArray();
-                    if (bytes.Length != file.Length || bytes.Length < 1)
-                    {
-                        _logger.Warning(
-                            "The data file for the transaction {TxId} seems corrupted; " +
-                            "it will be treated nonexistent and removed at all.",
-                            txid
-                        );
-                        _txLock.EnterWriteLock();
-                        try
-                        {
-                            _db.FileStorage.Delete(fileId);
-                        }
-                        finally
-                        {
-                            _txLock.ExitWriteLock();
-                        }
+            UPath path = TxPath(txid);
+            if (!_txs.FileExists(path))
+            {
+                return null;
+            }
 
-                        return null;
-                    }
-                }
+            byte[] bytes;
+            try
+            {
+                bytes = _txs.ReadAllBytes(path);
             }
-            finally
+            catch (FileNotFoundException)
             {
-                _txLock.ExitUpgradeableReadLock();
+                return null;
             }
 
-            return Transaction<T>.FromBencodex(bytes);
+            Transaction<T> tx = Transaction<T>.FromBencodex(bytes);
+            _txCache.AddOrUpdate(txid, tx);
+            return tx;
         }
 
         /// <inheritdoc/>
         public override void PutTransaction<T>(Transaction<T> tx)
         {
-            string fileId = TxFileId(tx.Id);
-            string filename = tx.Id.ToHex();
-            byte[] txBytes = tx.ToBencodex(true);
-            _txLock.EnterUpgradeableReadLock();
-            try
+            if (_txCache.ContainsKey(tx.Id))
             {
-                LiteFileInfo file = _db.FileStorage.FindById(fileId);
-                if (file is LiteFileInfo)
-                {
-                    // No-op if already exists.
-                    return;
-                }
+                return;
+            }
+
+            byte[] txBytes = tx.ToBencodex(true);
```

Comment:
How about checking the transaction not only from the cache but also from the file system?

Reply:
Fixed: feb6fca3c055c92fa3b8c233a3a071b457ecf8d8.